### PR TITLE
[PVR][Estuary] Feature: Persistent EPG searches.

### PR DESCRIPTION
**Pull request by ksooo**
## Description
This PR adds a feature to save PVR guide (EPG) searches.

## Motivation and context
I felt this could be useful and IIRC it was also requested in Kodi forum.

## How has this been tested?
Four weeks of extensive testing in my living room (Android) and my dev machine (macOS).

## What is the effect on users?
Ability to save, load, edit PVR guide changes.

## Screenshots (if appropriate):

screenshot00001

screenshot00004

screenshot00005

screenshot00006

screenshot00008

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/pvr/epg/EpgContainer.h [hunk 1]
```diff
@@ -225,6 +226,34 @@ namespace PVR
...
+    /*!
+     * @brief Get all saved searches from the database.
+     * @param bRadio Whether to fetch saved searches for radio or TV.
+     * @return The searches.
+     */
+    std::vector<std::shared_ptr<CPVREpgSearchFilter>> GetSavedSearches(bool bRadio);
+
+    /*!
+     * @brief Persist a saved search in the database.
+     * @param search The saved search.
```
Comment (AlwinEsch) [inline xbmc/pvr/epg/EpgContainer.h:238]: Maybe change to `@param[in,out] search The saved search.` to know why not const and to add a small note what becomes edited.

Found this as edit:
- `epgSearch.SetDatabaseId(static_cast<int>(m_pDS->lastinsertid()));`
- `epgSearch.SetChanged(false);`

And the read of CPVREpgSearchFilter values from it.

**Review by AlwinEsch — COMMENTED** [review]
Your code looks perfect, not seen any bad parts inside and a very nice feature will come with :+1: 

The only part was about the docu text, but thats all. Normally I like to use always a `@param[in], @param[out] or @param[in,out]` where makes easier in which ways it work.

File: xbmc/pvr/epg/EpgContainer.h (hunk 1 above)
Comment (ksooo) [reply xbmc/pvr/epg/EpgContainer.h:238]: Fair point, and if this pay for you I would prefer to change this throughout the whole pvr module in a follow up PR, not in this PR.

File: xbmc/FileItem.h [hunk 2]
```diff
@@ -294,6 +296,13 @@ class CFileItem :
     return m_epgInfoTag;
   }
 
+  inline bool HasEPGSearchFilter() const { return m_epgSearchFilter.get() != nullptr; }
```
Comment (howie-f) [inline xbmc/FileItem.h:299]: does this pointer need to be unwrapped by .get()? also i think compilers automatically assume `inline` for 1-liner funcs in header files… may be obsolete here, i know it‘s widely used in CFileItem.h

Comment (ksooo) [reply xbmc/FileItem.h:299]: > does this pointer need to be unwrapped by .get()? 

Sorry, what do you mean? What code change do you suggest?

Comment (howie-f) [reply xbmc/FileItem.h:299]: this :) `bool HasEPGSearchFilter() const { return m_epgSearchFilter != nullptr; }`

Comment (ksooo) [reply xbmc/FileItem.h:299]: Ah, yes. Makes perfect sense.

**Review by ronie — APPROVED** [review]
skin changes look ok to me.

File: xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp [hunk 3]
```diff
@@ -281,24 +329,49 @@ void CGUIDialogPVRGuideSearch::Update()
...
+    if (!m_startDateTime.IsValid())
+      m_startDateTime = CDateTime::GetUTCDateTime();
+  }
+
+  m_endDateTime = m_searchFilter->GetEndDateTime();
+  if (!m_endDateTime.IsValid())
+  {
+    m_endDateTime = CServiceBroker::GetPVRManager().EpgContainer().GetLastEPGDate();
+    if (!m_endDateTime.IsValid())
+      m_endDateTime = m_startDateTime + CDateTimeSpan(10, 0, 0, 0);
```
Comment (phunkyfish) [inline xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp:351]: How much time you are adding is non obvious from the code. It’s akin to passing many bool values!

**Review by phunkyfish — APPROVED** [review]
Really nice work. Just one minor nitpick is all. Up to you.

File: xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp (hunk 3 above)
Comment (ksooo) [reply xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp:351]: You mean a should add a comment? I thought everybody is familiar with CDateTimeSpin and knows the semantics of the four parameters. ;-) However, I will add a comment.

Comment (phunkyfish) [reply xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp:351]: I rarely interact with it myself. I generally leave the date stuff to others 😉

Comment (ksooo) [reply xbmc/pvr/dialogs/GUIDialogPVRGuideSearch.cpp]: done.

**Pull request merged**